### PR TITLE
Merge BaseHandle into Handle, reorganize Handle and Jdbi methods, add fluent returns

### PR DESCRIPTION
**Pull request by stevenschlansker**
We never really had any implementations of `Handle` other than `BasicHandle` so we can reduce our code by just removing it.

### CODE REVIEW

File: core/src/main/java/org/jdbi/v3/Handle.java [hunk 1]
```diff
@@ -28,109 +32,231 @@
...
+    }
 
     /**
-     * Start a transaction
+     * Register a column mapper which will have its parameterized type inspected to determine what it maps to.
      *
-     * @return the same handle
+     * Column mappers may be reused by {@link RowMapper} to map individual columns.
+     *
+     * @param mapper the column mapper
```
Comment (qualidafial) [inline core/src/main/java/org/jdbi/v3/Handle.java]: We should add a `@throws` here for the case that the mapper isn't concretely parameterized.

File: core/src/main/java/org/jdbi/v3/Handle.java [hunk 2]
```diff
@@ -28,109 +32,231 @@
...
+    }
 
     /**
-     * Rollback a transaction
+     * Register a row mapper which will have its parameterized type inspected to determine what it maps to.
      *
-     * @return the same handle
+     * Will be used with {@link Query#mapTo(Class)} for registered mappings.
+     *
+     * @param mapper the row mapper
```
Comment (qualidafial) [inline core/src/main/java/org/jdbi/v3/Handle.java]: Ditto for `@throws`

**Comment by stevenschlansker** [comment]
added the `UnsupportedOperationException` docs.

File: core/src/main/java/org/jdbi/v3/Handle.java [hunk 3]
```diff
@@ -28,109 +32,237 @@
...
+import org.jdbi.v3.tweak.TransactionHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * This represents a connection to the database system. It usually is a wrapper around
  * a JDBC Connection object.
  */
-public interface Handle extends Closeable
+public class Handle implements Closeable
```
Comment (qualidafial) [inline core/src/main/java/org/jdbi/v3/Handle.java]: Should we make this final for now?

Comment (stevenschlansker) [inline core/src/main/java/org/jdbi/v3/Handle.java]: I'm ambivalent, you can't subclass it anyway (no public or protected constructor).  So I'm not going to do it unless there's a compelling reason.

Comment (umcodemonkey) [inline core/src/main/java/org/jdbi/v3/Handle.java]: I would discourage it; making it final makes it harder to mock.

Comment (qualidafial) [inline core/src/main/java/org/jdbi/v3/Handle.java]: Let's leave it then.

File: core/src/main/java/org/jdbi/v3/Handle.java [hunk 4]
```diff
@@ -140,225 +272,337 @@
...
+        finally {
+            try {
+                setTransactionIsolation(initial);
+            }
+            catch (RuntimeException e) {
+                if (! failed) {
+                    throw e;
+                }
+
+                // Ignore, there was already an exceptional condition and we don't want to clobber it.
```
Comment (qualidafial) [inline core/src/main/java/org/jdbi/v3/Handle.java]: `addSuppressed` instead, maybe?

File: core/src/main/java/org/jdbi/v3/Jdbi.java [hunk 5]
```diff
@@ -163,16 +229,38 @@ public Jdbi installPlugin(JdbiPlugin plugin)
...
      * @param locator StatementLocator which will be used by all Handle instances
      *                created from this DBI
      */
-    public void setStatementLocator(StatementLocator locator)
+    public Jdbi setStatementLocator(StatementLocator locator)
     {
-        assert locator != null;
+        if (locator == null) {
+            throw new IllegalArgumentException("null statement locator");
+        }
```
Comment (qualidafial) [inline core/src/main/java/org/jdbi/v3/Jdbi.java]: Could alternatively use `Objects.requireNonNull(locator, "null statement locator")`

File: core/src/main/java/org/jdbi/v3/Jdbi.java [hunk 6]
```diff
@@ -278,8 +365,9 @@ public void registerRowMapper(RowMapperFactory factory) {
      *
      * @param mapper the column mapper
      */
-    public void registerColumnMapper(ColumnMapper<?> mapper) {
+    public Jdbi registerColumnMapper(ColumnMapper<?> mapper) {
```
Comment (qualidafial) [inline core/src/main/java/org/jdbi/v3/Jdbi.java]: This method wants a `@throws` too, similar to the one in `Handle`

File: core/src/main/java/org/jdbi/v3/Handle.java (hunk 4 above)
Comment (stevenschlansker) [inline core/src/main/java/org/jdbi/v3/Handle.java]: fixed with try-with-resources.

File: core/src/main/java/org/jdbi/v3/Jdbi.java [hunk 7]
```diff
@@ -289,8 +377,33 @@ public void registerColumnMapper(ColumnMapper<?> mapper) {
...
-    public void registerColumnMapper(ColumnMapperFactory factory) {
+    public Jdbi registerColumnMapper(ColumnMapperFactory factory) {
         config.mappingRegistry.addColumnMapper(factory);
+        return this;
+    }
+
+    /**
+     * Register a row mapper which will have its parameterized type inspected to determine what it maps to
+     *
+     * Will be used with {@link Query#mapTo(Class)} for registered mappings.
```
Comment (qualidafial) [inline core/src/main/java/org/jdbi/v3/Jdbi.java]: Do we want to also mention generated keys?

**Comment by qualidafial** [comment]
A few minor comments to consider, then :shipit:

File: core/src/main/java/org/jdbi/v3/Jdbi.java (hunk 5 above)
Comment (stevenschlansker) [inline core/src/main/java/org/jdbi/v3/Jdbi.java]: done

File: core/src/main/java/org/jdbi/v3/Jdbi.java (hunk 7 above)
Comment (stevenschlansker) [inline core/src/main/java/org/jdbi/v3/Jdbi.java]: maybe later, we should do a separate pass over the javadoc.  Trying to get the code cleaned up right now :)

**Pull request merged**